static/js/protein/data_processing_utils.ts

```diff
@@ -592,3 +530,328 @@ export function getProteinDescription(data: GraphNodes): string {
     return proteinDescription;
   }
 }
+
+/**
+ * Given array, key-maker function, and value-maker function, construct object from array with string-valued keys.
+ * This is essentially a TS implementation of a python dictionary comprehension.
+ *
+ * Unfortunately generic index signature parameters aren't supported by TS yet, so we enforce that keys are strings
+ * (since they're automatically converted to strings anyway).
+ *
+ * References:
+ *  - https://linuxtut.com/en/87ab3f313a75b547c278/
+ *  - https://stackoverflow.com/questions/13315131/enforcing-the-type-of-the-indexed-members-of-a-typescript-object
+ *
+ */
+export function objectFromArray<T, V>(
+  arr: T[],
+  keyFunc: (arrElt: T) => string,
+  valFunc: (arrElt: T) => V
+): { [key: string]: V } {
+  return arr.reduce(
+    (obj, elt) => Object.assign(obj, { [keyFunc(elt)]: valFunc(elt) }),
+    {}
+  );
+}
+
+/**
+ * Given an (unpacked) array of arrays, return its transpose (after trimming arrays to length of shortest one).
+ * This is essentially a TS implementation of python's built-in zip function.
+ *
+ * Adapted from implementation by Matt Mancini
+ *
+ * Thanks also to Carlos Israel Ortiz Garcia and Jack Davison for the below references:
+ *  - https://ncjamieson.com/typing-zipwith/#typing-zip
+ *  - https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-0.html#variadic-tuple-types
+ */
+// Recursive elementwise type inference on array of arrays
+type Zip<T> = T extends [Array<infer U>, ...infer Z] ? [U, ...Zip<Z>] : [];
+export function zip<T extends unknown[][]>(...arrays: T): Array<Zip<T>> {
+  const minLength = Math.min(...arrays.map((arr) => arr.length));
+  return Array.from(Array(minLength).keys()).map((_, i) => {
+    return arrays.map((array) => array[i]) as Zip<T>;
+  });
+}
+
+/**
+ * Convert DCID of form bio/<id> to id.
+ * Note: probably should be doing dcid: `bio/${string}`, but currently causes many type errors
+ */
+export function idFromDCID(dcid: string): string {
+  return dcid.replace("bio/", "");
+}
+
+/**
+ * Given id, convert to DCID of form bio/<id>
+ */
+export function dcidFromID(id: string): string {
+  return `bio/${id}`;
+}
+
+/**
+ * Given quantity DCID of the form "<quantityName><quantityValue>", extract and return quantityValue.
+ */
+export function quantityFromDCID(
+  quantityDCID: string,
+  quantityName: string
+): number {
+  if (!quantityDCID.includes(quantityName)) {
+    throw `DCID "${quantityDCID}" does not contain quantity name "${quantityName}"`;
+  }
+  return Number(_.last(quantityDCID.split(quantityName)));
+}
+
+/**
+ * Given interaction DCID of the form bio/{protein1}_{protein2}, return [protein1, protein2]
+ */
+export function proteinsFromInteractionDCID(interactionDCID: string): string[] {
```

Comment:
Do we get access to the underlying nodes for these? That can avoid the assumptions for the structure of these dcids.

Would be helpful to have an example in the comments here - e.g. bio/2A5D_YEAST_AHA1_YEAST

Reply:
I've asked Bo a similar question before and I think this would require another hop in the knowledge graph (which from the clientside would be another request).  Migrating to flask and using the [interactionProtein](https://datacommons.org/browser/bio/1433G_HUMAN_P53_HUMAN) property would be much cleaner than what I'm currently doing.  

This might be overkill but another request might not be too bad because it could be sent in parallel with the request for interaction scores. `interactinType` and `interactionDetectionMethod` might also be useful to pull. (but yes this would all be better in flask)